New topic structure for authenticated devices.

sysexcontrol opened the pull request:
The previous topic structure requested devices to know their logical
device-id. If they authenticate themselves this is not necessary, so
they can use either the short form "/telemetry" or the fully qualified
topic address "/telemetry/tenantId/[auth-id|device-id]".

Signed-off-by: Karsten Frank <Karsten.Frank@bosch-si.com>

sophokles73 commented:
As I mentioned on the mailing list, FMPOV a device **never** publishes to `/telemetry/tenant/authId` but only to `/telemetry/tenant/deviceId`. If the device has been authenticated, it MAY publish to just `/telemetry`, but if it chooses to publish to the FQTA then the topic MUST contain the device ID only.
NB The auth ID that the device uses for authentication purposes MAY have the same value as its device ID. But in any case, the tupels (tenant, device ID) and (tenant, auth ID) MUST be unique.

Based on this, IMHO this PR should be closed.

sysexcontrol commented:
That is a good outcome of all the discussion, I will change this PR and provide an update. I am sure that soon afterwards we can merge it.
- no logicalDeviceId anymore
- no authId in FQTAs

sysexcontrol commented:
Provided update - eliminated all occurences of "logical device id" and changed it to "device-id", changed `telemetry/tenant-id/auth-id` back to `telemetry/tenant-id/device-id`.
Short review welcome, IMHO it can be merged soon - WDYT?

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -288,17 +288,21 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
     void handleCredentialsResult(final AsyncResult<String> attempt, final MqttEndpoint endpoint, final UsernamePasswordCredentials authObject) {
         if (attempt.succeeded()) {
-            String logicalDeviceId = attempt.result();
+            final String logicalDeviceId = attempt.result();
```
this should be just `deviceId` as well, I guess

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -288,17 +288,21 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
             endpoint.publishHandler(message -> {
                 final ResourceIdentifier resource = ResourceIdentifier.fromString(message.topicName());
 
+                // check that the device publishes to it's authId
```
to its *device-id*

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -288,17 +288,21 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
                 final ResourceIdentifier resource = ResourceIdentifier.fromString(message.topicName());
 
+                // check that the device publishes to it's authId
+                // TODO: should we support the logicalId as well in last part?
```
this comment can be removed IMHO

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -288,17 +288,21 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
                     // if MQTT client does not conform to the topic structure, close the connection (MQTT has no way for errors)
                     endpoint.close();
                 } else {
+                    // the payload is ALWAYS published to the logicalDeviceId downstream
```
device-id

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -340,21 +344,13 @@ private void publishMessage(final MqttEndpoint endpoint, final String tenantId,
-            } else {
-                Future<String> assertionTracker = getRegistrationAssertion(endpoint, tenantId, logicalDeviceId);
-                Future<MessageSender> senderTracker = getSenderTracker(message, resource, tenantId);
+            Future<String> assertionTracker = getRegistrationAssertion(endpoint, tenantId, logicalDeviceId);
```
`deviceId`

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -340,21 +344,13 @@ private void publishMessage(final MqttEndpoint endpoint, final String tenantId,
-                }, messageTracker);
-            }
+            CompositeFuture.all(assertionTracker, senderTracker).compose(ok -> {
+                doUploadMessage(logicalDeviceId, assertionTracker.result(), endpoint, message,
```
`deviceId`

sophokles73 commented on site/content/component/mqtt-adapter.md:
```diff
@@ -201,30 +201,50 @@ The same holds true analogously for the *hono-service-device-registry.hono* addr
-* Topic: `telemetry/${tenantId}/${deviceId}`
-* Client-id: ${deviceId}
+* Topic: `telemetry/${tenant-id}/${device-id}` or just `telemetry` (short form)
+* Client-id: not specified, use e.g. ${device-id} or any unique identifier 
```
just remove this line, otherwise people might think they need to set it explicitly while most (if not all) MQTT clients will set this to a generated value if not set

sophokles73 commented on site/content/component/mqtt-adapter.md:
```diff
@@ -201,30 +201,50 @@ The same holds true analogously for the *hono-service-device-registry.hono* addr
-Upload a JSON string for device `4711`:
+Upload a JSON string for device-id `4711`:
 
     $ mosquitto_pub -i 4711 -u 'sensor1@DEFAULT_TENANT' -P hono-secret -t telemetry/DEFAULT_TENANT/4711 -m '{"temp": 5}'
```
remove the `-i` param (the client ID)

sophokles73 commented on site/content/component/mqtt-adapter.md:
```diff
@@ -201,30 +201,50 @@ The same holds true analogously for the *hono-service-device-registry.hono* addr
 
+or (if the device does not know the device-id itself)
+
+    $ mosquitto_pub -i sensor1 -u 'sensor1@DEFAULT_TENANT' -P hono-secret -t telemetry -m '{"temp": 5}'
```
remove the `-i` param

sophokles73 commented on site/content/component/mqtt-adapter.md:
```diff
@@ -201,30 +201,50 @@ The same holds true analogously for the *hono-service-device-registry.hono* addr
+### Upload Telemetry Data (unauthenticated devices)
+
+* Topic: `telemetry/${tenant-id}/${device-id}`
+* Client-id: not specified, use e.g. ${device-id} or any unique identifier 
```
remove line

sophokles73 commented on site/content/component/mqtt-adapter.md:
```diff
@@ -201,30 +201,50 @@ The same holds true analogously for the *hono-service-device-registry.hono* addr
+
+Upload a JSON string for device-id `4711`:
+
+    $ mosquitto_pub -i 4711 -t telemetry/DEFAULT_TENANT/4711 -m '{"temp": 5}'
```
remove `-i` param

sophokles73 commented on site/content/component/mqtt-adapter.md:
```diff
@@ -201,30 +201,50 @@ The same holds true analogously for the *hono-service-device-registry.hono* addr
-* Topic: `event/${tenantId}/${deviceId}`
-* Client-id: ${deviceId}
+* Topic: `event/${tenant-id}/${device-id}` or just `event` (short form)
+* Client-id: not specified, use e.g. ${device-id} or any unique identifier 
```
remove line

sophokles73 commented on site/content/component/mqtt-adapter.md:
```diff
@@ -233,9 +253,22 @@ Upload a JSON string for device `4711`:
+    
+or (if the device does not know the device-id itself)
+
+    $ mosquitto_pub -i sensor1 -u 'sensor1@DEFAULT_TENANT' -P hono-secret -t event -m '{"alarm": 1}'
```
remove `-i` param

sophokles73 commented on site/content/component/mqtt-adapter.md:
```diff
@@ -233,9 +253,22 @@ Upload a JSON string for device `4711`:
+### Send Event Message (unauthenticated devices)
+
+* Topic: `event/${tenant-id}/${device-id}`
+* Client-id: not specified, use e.g. ${device-id} or any unique identifier 
```
remove line

sophokles73 commented on site/content/component/mqtt-adapter.md:
```diff
@@ -233,9 +253,22 @@ Upload a JSON string for device `4711`:
+
+Upload a JSON string for device `4711`:
+
+    $ mosquitto_pub -i 4711 -t event/DEFAULT_TENANT/4711 -m '{"alarm": 1}'
```
remove `-i` param

ppatierno commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -288,17 +288,21 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
         if (attempt.succeeded()) {
-            String logicalDeviceId = attempt.result();
+            final String logicalDeviceId = attempt.result();
             LOG.debug("successfully authenticated device id [{}]", logicalDeviceId);
```
deviceId here

ppatierno commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -288,17 +288,21 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
 
+                // check that the device publishes to it's authId
+                // TODO: should we support the logicalId as well in last part?
                 if (!validateCredentialsWithTopicStructure(resource, authObject.getTenantId(), logicalDeviceId)) {
```
deviceId

sysexcontrol commented:
@sophokles73 : thanks for the comments, all +1, already fixed them.

@ppatierno ; Now taking your comments into account, then proiding an update...

sysexcontrol commented:
@ppatierno : regarding the clientId in log messages:
I agree, but maybe we should keep it for the start and the end of a MqttEndpoint connection.
Otherwise there is nothing else to log that could identify the client in this lifecycle phase. 
A log like "some client has connected" is not worth much anymore. 
So, either still log the clientId, or delete these kind of log messages completely - WDYT?

Will eliminate all other occurences of clientId in logs in any case.

ppatierno commented:
@sysexcontrol I agree with you but then we should refuse a connect with an empty client-id ? It's allowed by the MQTT 3.1.1 spec (but clean-session MUST be true).
If empty client-id is allowed and used for logging we don't have information about that.
@sophokles73 asked to delete the -i parameter from mosquitto_pub which for the example is good because mosquitto_pub generates a random client-id ... but what for other clients ?

sysexcontrol commented:
@ppatierno :
allow empty clientId : would it be so bad if we have nothing to log then? Meaning: if only the empty log message is bothering us, I wouldn't refuse this for only that reason. 
Maybe we can keep the decision about empty clientId out this PR to get it finished soon?

delete the "-i" in the examples: I agree with @sophokles73 that the "-i" might create more confusion than it would help anyone, so better delete it completely.
And if we do not forbid empty clientIds (of other clients) this would not confuse users of other clients either - WDYT?

ppatierno commented:
@sysexcontrol I agree on deleting -i and keep decision about empty clientid out of this PR. I want it soon for testing and having a working environment :-)

sysexcontrol commented:
@sophokles73 @ppatierno : pushed latest update, all comments of you should be considered now.

ppatierno commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -288,78 +288,75 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
             endpoint.closeHandler(v -> {
-                LOG.debug("connection closed with client [{}], authId [{}], deviceId [{}]",
-                        endpoint.clientIdentifier(), authObject.getAuthId(), logicalDeviceId);
+                LOG.debug("connection closed with tenantId [{}], authId [{}], deviceId [{}]",
```
maybe something like "connection closed with client (tenantId [{}], authId [{}], deviceId [{}])" ?

sophokles73 replied:
`closed connection to device [tenant-id: {}, auth-id: {}, device-id: {}]`

ppatierno commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -288,78 +288,75 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
         } else {
-            LOG.debug("authentication failed for client [{}] rejected [{}]",
-                    endpoint.clientIdentifier(), MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED);
+            LOG.debug("authentication failed for tenantId [{}], authId [{}] rejected [{}]",
```
Ditto as above

ppatierno commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -288,78 +288,75 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
             // MQTT client is trying to publish on invalid topic; it does not contain at least two segments
-            LOG.debug("client [ID: {}] tries to publish on unsupported topic", endpoint.clientIdentifier());
+            LOG.debug("client [ID: {}] for [tenantId: [{}], deviceId: [{}] tries to publish on unsupported topic",
+                    endpoint.clientIdentifier(), tenantId, deviceId);
```
still using client-id here ?

sysexcontrol replied:
Oh, this wasn't intended! :+1:

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -248,11 +248,11 @@ private void handleEndpointConnectionWithoutAuthentication(final MqttEndpoint en
             }
             if (resource.getTenantId() == null) {
-                // if MQTT client doesn't specify device_id then closing connection (MQTT has no way for errors)
+                // if MQTT client doesn't specify device-id then closing connection (MQTT has no way for errors)
```
tenant ID in this case

sysexcontrol replied:
thx!

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -288,78 +288,75 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
+    private void publishMessage(final MqttEndpoint endpoint, final String tenantId, final String deviceId,
             final MqttPublishMessage message, final ResourceIdentifier resource) {
-        LOG.trace("received message [client ID: {}, topic: {}, QoS: {}, payload {}]", endpoint.clientIdentifier(),
+        LOG.trace("received message [tenantId [{}], deviceId [{}], topic: [{}], QoS: [{}], payload [{}]]", tenantId, deviceId,
```
no need to nest brackets here, i.e.
`received message [tenant-id: {}, device-id: {}, ... ]` will do fine.

sysexcontrol replied:
would keep the brackets around payload though (could be quite awful to read otherwise?)

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -288,78 +288,75 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
                 if (s.failed()) {
-                    LOG.debug("cannot process message [client ID: {}, deviceId: {}, topic: {}, QoS: {}]: {}",
-                            endpoint.clientIdentifier(), logicalDeviceId,
+                    LOG.debug("cannot process message [tenantId: [{}], deviceId: [{}], topic: [{}], QoS: [{}]]: {}",
```
same here

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -288,78 +288,75 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
                 } else {
-                    LOG.trace("successfully processed message [client ID: {}, deviceId: {}, topic: {}, QoS: {}]",
-                            endpoint.clientIdentifier(), logicalDeviceId,
+                    LOG.trace("successfully processed message [tenantId: [{}], deviceId: [{}], topic: [{}], QoS: [{}]]",
```
and here

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -288,78 +288,75 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
 
             // MQTT client is trying to publish on invalid topic; it does not contain at least two segments
-            LOG.debug("client [ID: {}] tries to publish on unsupported topic", endpoint.clientIdentifier());
+            LOG.debug("client [ID: {}] for [tenantId: [{}], deviceId: [{}] tries to publish on unsupported topic",
```
and here

sophokles73 commented on site/content/component/mqtt-adapter.md:
```diff
@@ -201,41 +201,70 @@ The same holds true analogously for the *hono-service-device-registry.hono* addr
+or (if the device does not know the device-id itself)
 
-    $ mosquitto_pub -i 4711 -u 'sensor1@DEFAULT_TENANT' -P hono-secret -t telemetry/DEFAULT_TENANT/4711 -m '{"temp": 5}'
+    $ mosquitto_pub -u 'sensor1@DEFAULT_TENANT' -P hono-secret -t telemetry -m '{"temp": 5}'
```
IMHO we should use the short version (`telemetry` topic) as the example given and mention the FQTA version in a note only.

sysexcontrol replied:
@sophokles73 : I agree to put the short version to the top, but we have a lot of disagreement regarding this point.
So I probably should wait for more comments regarding this?
@ppatierno @pellmann : can you please post a short comment to this point also? Thanks!

ppatierno replied:
after long discussion on chat we agreed on not having "auth-id" in the topic. I'm ok with that because it's part of the credentials which should not appear in the topic. Then the device could not know its device-id (former logical-id) so the only topic format we can use is the simple one "telemetry".

pellmann replied:
For me the point of discussion here was if we need to separate between the device-id and a "logical-id" (solution-device-id) that is not known by the device. I thought it would be easier without this separation and clearer to always use the FQTA version - BUT if you want to stay with it, I think the short version is the natural one.

sysexcontrol replied:
Thanks to all - I read this as an agreement that I now put the STA version as main example and mention the FQTA as a note then.

sophokles73 commented on site/content/component/mqtt-adapter.md:
```diff
@@ -201,41 +201,70 @@ The same holds true analogously for the *hono-service-device-registry.hono* addr
+    
+or (if the device does not know the device-id itself)
+
+    $ mosquitto_pub -u 'sensor1@DEFAULT_TENANT' -P hono-secret -t event -m '{"alarm": 1}'
```
same here: use short version and add note for long version

sysexcontrol commented:
Pushed update - no brackets anymore except around clientId and payload,
other comments considered except the order of the examples of "telemetry" vs. "telemetry/tenantId/deviceId" (see direct comment to this idea requesting for more voices to hear)...

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -307,37 +307,37 @@ void handleCredentialsResult(final AsyncResult<String> attempt, final MqttEndpoi
     private void publishMessage(final MqttEndpoint endpoint, final String tenantId, final String deviceId,
             final MqttPublishMessage message, final ResourceIdentifier resource) {
-        LOG.trace("received message [tenantId [{}], deviceId [{}], topic: [{}], QoS: [{}], payload [{}]]", tenantId, deviceId,
+        LOG.trace("received message for device (tenantId: {}, deviceId: {}, topic: {}, QoS: {}, payload: [{}])", tenantId, deviceId,
```
I think we cannot assume the payload to be a string here. In any case, we should at least put this log statement into an if  block that is only executed if the logger is actually configured for trace level. Otherwise we always try to unmarshal the payload into a string, regardless of whether it actually gets logged or not. IMHO we should not log the payload anyway, because we might also open up a door for attackers to actually wreck our log ... the payload size should be enough FMPOV.

sysexcontrol replied:
Full ack: I will remove the payload log.

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -307,37 +307,37 @@ void handleCredentialsResult(final AsyncResult<String> attempt, final MqttEndpoi
             messageTracker.setHandler(s -> {
                 if (s.failed()) {
-                    LOG.debug("cannot process message [tenantId: [{}], deviceId: [{}], topic: [{}], QoS: [{}]]: {}",
+                    LOG.debug("cannot process message for device (tenantId: {}, deviceId: {}, topic: {}, QoS: {}]: {})",
```
wrong type of opening bracket

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -355,8 +355,8 @@ private void publishMessage(final MqttEndpoint endpoint, final String tenantId,
             // MQTT client is trying to publish on invalid topic; it does not contain at least two segments
-            LOG.debug("client [ID: {}] for [tenantId: [{}], deviceId: [{}] tries to publish on unsupported topic",
-                    endpoint.clientIdentifier(), tenantId, deviceId);
+            LOG.debug("client for (tenantId: {}, deviceId: {}) tries to publish on unsupported topic",
```
wrong type of brackets

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -307,37 +307,37 @@ void handleCredentialsResult(final AsyncResult<String> attempt, final MqttEndpoi
                     close(endpoint);
                 } else {
-                    LOG.trace("successfully processed message [tenantId: [{}], deviceId: [{}], topic: [{}], QoS: [{}]]",
+                    LOG.trace("successfully processed message for device (tenantId: {}, deviceId: {}, topic: {}, QoS: {})",
```
wrong type of brackets

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -307,37 +307,37 @@ void handleCredentialsResult(final AsyncResult<String> attempt, final MqttEndpoi
 
         } else {
-            LOG.debug("authentication failed for tenantId [{}], authId [{}] rejected [{}]",
+            LOG.debug("authentication failed for device (tenantId: {}, authId: {} rejected: {})",
```
wong type of brackets

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -307,37 +307,37 @@ void handleCredentialsResult(final AsyncResult<String> attempt, final MqttEndpoi
 
             endpoint.closeHandler(v -> {
-                LOG.debug("connection closed with tenantId [{}], authId [{}], deviceId [{}]",
+                LOG.debug("connection closed to device (tenantId: {}, authId: {}, deviceId: {})",
```
wrong type of brackets

ppatierno commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -388,18 +385,19 @@ private void publishMessage(final MqttEndpoint endpoint, final String tenantId,
                 // the registration assertion has been returned, avoid to put it into the map
                 if (endpoint.isConnected()) {
-                    LOG.trace("caching registration assertion for client [{}]", endpoint.clientIdentifier());
+                    LOG.trace("caching registration assertion for [tenantId: [{}], deviceId: [{}]",
```
There is a "[" bracket not closed at the end but are we using "(" brackets around such logs ?

sophokles73 replied:
No, we are using `[.....]` in logs

ppatierno replied:
But having something like "]]" at the end isn't so cool :-) 
I agree on using "[ .. ]" on fields values for example (i.e. tenantId, deviceId, ...) but for representing an "object", wrapping it into different brackets could be more readable.

sophokles73 replied:
That's why I asked Karsten to remove the nesting so it reads
`LOG.trace("caching registration assertion for [tenantId: {}, deviceId: {}]" ...`

Done

sysexcontrol commented:
Update coming soon...
The "(" ")" brackets were from a previous comment from @ppatierno, completely removed them and only use "[" "]" now...
In a few minutes I push the update.

sysexcontrol commented:
Update avail.

sophokles73 commented on adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java:
```diff
@@ -288,78 +288,74 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
         } else {
-            LOG.debug("authentication failed for client [{}] rejected [{}]",
-                    endpoint.clientIdentifier(), MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED);
+            LOG.debug("authentication failed for device [tenantId: {}, authId: {} rejected: {}]",
```
I guess the closing bracket is supposed to be after `authId: {}` in this case, right? Otherwise there is a comma missing and I would replace *rejected* with *error code* or *cause* ...
